Compfiles/Imo1970P4.lean

```diff
@@ -21,15 +20,831 @@ equals the product of the nubmers in the other set.
 
 namespace Imo1970P4
 
-determine SolutionSet : Finset ℕ+ := sorry
+snip begin
+
+lemma card_opposite (s s' s'' : Finset ℕ) (predicate: ℕ → Prop) [DecidablePred predicate] (filter : s' = (s.filter (λ x => predicate x)))
+                    (opposite_filter: s'' = (s.filter (λ x => ¬ predicate x))) : s'.card + s''.card = s.card := by
+  rw[filter]
+  rw[opposite_filter]
+  have := @Finset.filter_card_add_filter_neg_card_eq_card ℕ s predicate
+  apply this
+
+lemma prime_dvd_elem_of (p : ℕ) (pp : Nat.Prime p) (s : Finset ℕ) : p ∣ (∏ m ∈ s, m) → ∃ x ∈ s, p ∣ x := by
+  exact Prime.exists_mem_finset_dvd pp.prime
+
+lemma no_other_p_divisors_nearby (x : ℕ) (y : ℕ) (p : ℕ) (p_gt_5 : p > 5) (x_lt_y : x < y) (close_by: ∃ k, k ≤ 5 ∧ x + k = y) (x_div_p : p ∣ x) : ¬ (p ∣ y) := by
+  obtain ⟨k, ⟨bound, sum⟩⟩ := close_by
+  intro H
+  obtain ⟨a, Ha⟩ := x_div_p
+  obtain ⟨b, Hb⟩ := H
+  rw[Ha] at sum
+  rw[Hb] at sum
+  rw[Ha] at x_lt_y
+  rw[Hb] at x_lt_y
+  have a_lt_b : a < b := by
+    exact (mul_lt_mul_left (show 0 < p by omega)).mp x_lt_y
+  have a_lt_b_2 : 1 ≤ (b - a) := by
+    omega
+  have k_eq : p * (b - a) = k := by
+    calc p * (b - a) = p * b - p * a := mul_tsub p b a
+    _ = p * a + k - p * a := by rw[sum]
+    _ = k := by omega
+  have : p * (b - a) > 5 := by
+    calc p * (b - a) > 5 * (b - a) := by rel[p_gt_5]
+         _ ≥ 5 * 1 := by rel[a_lt_b_2]
+  have : k > 5 := by
+    omega
+  omega
+
+lemma no_other_5_divisors_nearby (x : ℕ) (y : ℕ) (x_lt_y : x < y) (close_by: ∃ k, k ≤ 4 ∧ x + k = y) (x_div_p : 5 ∣ x) : ¬ (5 ∣ y) := by
+  omega
+
+-- The next few functions apply the following logic from the Art of Problem Solving Solution 1
+
+-- There are exactly three odd numbers in s
+-- p > 5 is not a divisor of any of those three odd numbers
+-- p = 5 is a divisor of at most one of those three odd numbers
+-- p = 3 is a divisor of exactly one of those three odd numbers
+-- because these numbers are odd, p = 2 is not a divisor of any of those odd numbers
+-- p = 2, p = 3, p = 5 and p > 5 are all the prime numbers out there.
+-- by pigeonhole principle, one odd number must not have any prime divisors
+-- this number must be 1.
+-- The only two intervals that contain 1 are {1, 2, 3, 4, 5, 6} and {0, 1, 2, 3, 4, 5}
+
+
+
+lemma elems_in_interval_nearby (x y n : ℕ ) (s : Finset ℕ) (interval : s = Finset.Icc n (n + 5))
+  (x_in_s : x ∈ s) (y_in_s : y ∈ s) (x_lt_y : x < y) : ∃ k ≤ 5, x + k = y := by
+  simp_all only [Finset.mem_Icc]
+  use y - x
+  constructor
```

Comment:
Don't need `constructor` -- `omega` works already.

Reply:
Fixing in the next patch